src/onehot.cpp

```diff
@@ -0,0 +1,92 @@
+/*******************************************************************************
+ *
+ * MIT License
+ *
+ * Copyright (c) 2024 Advanced Micro Devices, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+
+#include <miopen/datatype.hpp>
+#include <miopen/find_solution.hpp>
+#include <miopen/float_equal.hpp>
+#include <miopen/kernel_cache.hpp>
+#include <miopen/onehot/solvers.hpp>
+#include <miopen/onehot/invoke_params.hpp>
+#include <miopen/onehot.hpp>
+#include <miopen/tensor.hpp>
+
+namespace miopen {
+
+miopenStatus_t OneHot(Handle& handle,
+                      TensorDescriptor& inDesc,
+                      const void* input,
+                      long inputSize,
+                      TensorDescriptor& outDesc,
+                      void* output,
+                      int numClasses)
+{
+    // If num classes is to -1(default), the number of classes will be inferred as one greater than
+    // the largest class value in the input tensor.
+    if(numClasses == -1)
+    {
+        for(auto i = 0; i < inputSize; ++i)
+        {
+            numClasses = std::max(numClasses, static_cast<const int*>(input)[i] + 1);
+        }
+    }
+
+    const int* input_int = static_cast<const int*>(input);
```

Comment:
I'm not sure event if you can check this on the host side.
I think it's better to check in the kernel, as the original modnn implementation does.
@seungmanhan @et16kr Any opinion on this?

Reply:
Since the values pointed to by input are in GPU memory, I have doubts about whether this will work.
There was a similar case with getitem. I created an error tensor then pulled the kernel output to the CPU so the user could figure out if there was a problem.

Reply:
@seungmanhan Could you show me the location of the `getitem` case you talk about. I can't find it.

Reply:
Sorry the answer was late. https://github.com/ROCm/MIOpen/pull/2883

Reply:
I have added a tensor err param to the kernel function and change the test and driver to check the err code as follow:
```cpp
// kernel
extern "C" __global__ void OneHotContiguous( const INPUT_TYPE* input, OUTPUT_TYPE* output, ERR_TYPE* err, long input_size, int num_classes)
{
    oneHotContiguousKernel<INPUT_TYPE, OUTPUT_TYPE, ERR_TYPE>(
        input, output, err, input_size, num_classes);
}
```

```cpp
// test
void RunTest() {
   status = miopen::OneHot(...);
   EXPECT_EQ(status, miopenStatusSuccess);

   err.data = handle.Read<T>(err_dev, err.data.size());
   EXPECT_EQ(err.data[0], 0) << "Error: input tensor value is negative or greater than num_classes";
}
```

```cpp
// driver
int RunForwardGPU() {
    miopenOneHot(...)
    ...
    if(err.data()[0] == 1)
    {
        MIOPEN_THROW(miopenStatusBadParm, "Error: input tensor must not contains negative value.");
    }
    if(err.data()[0] == 2)
    {
        MIOPEN_THROW(miopenStatusBadParm,
                     "Error: input tensor value must be less than num_classes");
    }

    return miopenStatusSuccess;
}
```

Reply:
It looks okay. As a next work, add the ERROR_CODE_LARGER_THAN_NUM_CLASS macro to miopen.h and also add a result comparison for error in the driver results. Also works in the same way for gtest